File: libs/server/domain/src/domain/diagram/modeling.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::domain::{EntityAttribute, LogicalEntityType, Ref};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelingProposal {
    pub summary: String,
    pub changes: ModelingProposalChanges,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelingProposalChanges {
    #[serde(default)]
    pub add_entities: Vec<ModelingDraftEntity>,
    #[serde(default)]
    pub update_entities: Vec<ModelingDraftEntity>,
    #[serde(default)]
    pub delete_entities: Vec<String>,
    #[serde(default)]
    pub add_relationships: Vec<ModelingDraftRelationship>,
    #[serde(default)]
    pub update_relationships: Vec<ModelingDraftRelationship>,
    #[serde(default)]
    pub delete_relationships: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelingDraftEntity {
    pub id: String,
    pub name: String,
    pub label: Option<String>,
    #[serde(rename = "type")]
    pub entity_type: LogicalEntityType,
    pub sub_type: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub attributes: Vec<EntityAttribute>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelingDraftRelationship {
    pub id: Option<String>,
    pub source: Ref<String>,
    pub target: Ref<String>,
    pub label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum ModelingEvent {
    TextChunk {
        chunk: String,
    },
    ReasoningStarted,
    ReasoningChunk {
        chunk: String,
    },
    ReasoningEnded,
    ToolCallStarted {
        tool_call_id: String,
        tool_name: Option<String>,
    },
    ToolCallDelta {
        tool_call_id: String,
        tool_name: Option<String>,
        chunk: String,
    },
    ToolCallReady {
        tool_call_id: String,
        tool_name: String,
        input: Value,
    },
    ToolExecutionStarted {
        tool_call_id: String,
        tool_name: String,
        args: Value,
    },
    ToolExecutionUpdated {
        tool_call_id: String,
        tool_name: String,
        args: Value,
        partial_result: Value,
    },
    ToolExecutionEnded {
        tool_call_id: String,
        tool_name: String,
        result: Value,
        is_error: bool,
    },
    MessageEnded,
    AgentEnded,
    Completed,
}

File: libs/server/domain/src/domain/diagram/modeling.ts
import { EntityAttribute, LogicalEntityType } from '../logical-entity';
import { Ref } from '../core';

export interface ModelingProposal {
  summary: string;
  changes: ModelingProposalChanges;
}

export interface ModelingProposalChanges {
  addEntities: ModelingDraftEntity[];
  updateEntities: ModelingDraftEntity[];
  deleteEntities: string[];
  addRelationships: ModelingDraftRelationship[];
  updateRelationships: ModelingDraftRelationship[];
  deleteRelationships: string[];
}

export interface ModelingDraftEntity {
  id: string;
  name: string;
  label: string | null;
  type: LogicalEntityType;
  subType: string | null;
  description: string | null;
  attributes: EntityAttribute[];
}

export interface ModelingDraftRelationship {
  id: string | null;
  source: Ref<string>;
  target: Ref<string>;
  label: string | null;
}

export interface ModelingRequest {
  requirement: string;
  modelDirectory: string;
  signal?: AbortSignal;
}

export type ModelingEvent =
  | { type: 'text-chunk'; chunk: string }
  | { type: 'reasoning-started' }
  | { type: 'reasoning-chunk'; chunk: string }
  | { type: 'reasoning-ended' }
  | { type: 'tool-call-started'; toolCallId: string; toolName: string | null }
  | {
      type: 'tool-call-delta';
      toolCallId: string;
      toolName: string | null;
      chunk: string;
    }
  | {
      type: 'tool-call-ready';
      toolCallId: string;
      toolName: string;
      input: unknown;
    }
  | {
      type: 'tool-execution-started';
      toolCallId: string;
      toolName: string;
      args: unknown;
    }
  | {
      type: 'tool-execution-updated';
      toolCallId: string;
      toolName: string;
      args: unknown;
      partialResult: unknown;
    }
  | {
      type: 'tool-execution-ended';
      toolCallId: string;
      toolName: string;
      result: unknown;
      isError: boolean;
    }
  | { type: 'message-ended' }
  | { type: 'agent-ended' }
  | { type: 'completed' };

export interface DomainArchitect {
  proposeModelStream(request: ModelingRequest): AsyncIterable<ModelingEvent>;
}

export const DOMAIN_ARCHITECT = Symbol('DOMAIN_ARCHITECT');
